### PR TITLE
Merge latest bits from `main` branch into extensions branch

### PR DESCRIPTION
**Pull request by jcouv**
The last ~~two~~ three commits resolve conflicts.

### CODE REVIEW

**Comment by jcouv** [comment]
@AlekseyTs @jjonescz for review. Thanks

**Comment by jjonescz** [comment]
Looks like CI run failed: https://dev.azure.com/dnceng-public/public/_build/results?buildId=614874&view=results

**Comment by jcouv** [comment]
Thanks for the pointer! I'm going to do a new merge from scratch and will force-push, so that the commits for resolving conflicts remain at the end.

File: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs [hunk 1]
```diff
@@ -21571,7 +21571,7 @@ public static void M(dynamic d)
         var model = comp.GetSemanticModel(tree);
         var memberAccess = GetSyntax<ElementAccessExpressionSyntax>(tree, "(new C())[d]");
         Assert.Equal("System.Int32 C.this[System.Int32 i] { get; }", model.GetSymbolInfo(memberAccess).Symbol.ToTestDisplayString());
-        Assert.Equal(["System.Int32 C.this[System.Int32 i] { get; }"], model.GetMemberGroup(memberAccess).ToTestDisplayStrings());
+        Assert.Equal([], model.GetMemberGroup(memberAccess).ToTestDisplayStrings());
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:21574]: Why has this changed?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1083)

File: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:21574]: That's the way this test without the extension now behaves in `main` branch where some changes relative to `dynamic` were made.
Given that this scenario doesn't actually involve the extension member, I didn't investigate further.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs [hunk 2]
```diff
@@ -131,7 +131,7 @@ public override Conversion GetMethodGroupDelegateConversion(BoundMethodGroup sou
                     return false;
                 }
 
-                MethodGroupResolution resolution = _binder.ResolveMethodGroup(source, analyzedArguments: null, isMethodGroupConversion: false, ref useSiteInfo);
+                MethodGroupResolution resolution = _binder.ResolveMethodGroup(source, analyzedArguments: null, ref useSiteInfo, options: OverloadResolution.Options.IsMethodGroupConversion);
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs]: Mistake on my part

File: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:21574]: > That's the way this test without the extension now behaves in `main` branch where some changes relative to `dynamic` were made. Given that this scenario doesn't actually involve the extension member, I didn't investigate further.

The behavior looks consistent with early bound case (no `dynamic` involved):
```
        [Fact]
        public void Test()
        {
            var source = """
C.M(1);
class C
{
    public static void M(int d)
    {
        System.Console.Write((new C())[d]);
    }
    public int this[int i] => 42;
}
""";
            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
            CompileAndVerify(comp, expectedOutput: ("42"), verify: Verification.FailsPEVerify).VerifyDiagnostics();
            var tree = comp.SyntaxTrees.First();
            var model = comp.GetSemanticModel(tree);
            var memberAccess = GetSyntax<ElementAccessExpressionSyntax>(tree, "(new C())[d]");
            Assert.Equal("System.Int32 C.this[System.Int32 i] { get; }", model.GetSymbolInfo(memberAccess).Symbol.ToTestDisplayString());
            Assert.Equal([], model.GetMemberGroup(memberAccess).ToTestDisplayStrings());
        }
```

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 1084)

**Review by jjonescz — APPROVED** [review]

**Pull request merged**